### PR TITLE
Adding the IREE CPU task scheduler (iree/task/).

### PR DESCRIPTION
**Pull request by benvanik**
This is designed to be usable standalone and independent of the HAL. Future PRs will wire this up to the host HAL backend code strictly consuming its public API.

This initial PR has a feature staging demo in iree/task/executor_test.cc that I've been using during development to verify/benchmark features. It shows off how the system is used and when the HAL is wired up will be replaced with tests.

As a size reference, **the entire iree/task/ path contributes only ~10KB to an x64 binary** and has near zero dependencies (just ~2.5KB from iree/base/ for threading/synchronization/atomics, which would be amortized across other parts of the runtime). For comparison absl::Mutex *alone* in the same size-optimized LTO'ed build is 20KB (excluding other parts of absl and STL it pulls in). My goal here was to build something that gave us superpowers without any overhead and I think I nailed it :P Note also that this will replace the current iree/hal/host/ threading code that uses std::thread/absl::Mutex/etc so we'll be at a net-negative binary size change while also gaining a fine-grained task-based scheduling system 🤯

### CODE REVIEW

File: iree/task/executor_test.cc
Comment (ScottTodd) [inline iree/task/executor_test.cc:106]: Useful to keep these comments, `#if 1`s, etc.?

File: iree/task/task.c
Comment (ScottTodd) [inline iree/task/task.c:403]: relating to one of my other comments about a max of 64 workers - I'm wondering how much impact this sort of topological scheduling will matter on consumer systems

File: iree/task/task.c
Comment (ScottTodd) [inline iree/task/task.c:653]: Intended to be uncommented, given the performance note?

I'm also wondering at a glance if the x/y/z are meant to be appended in the inner loop instead of up a level. Could use a bit more explanation in a comment.

File: iree/task/task.c
Comment (ScottTodd) [inline iree/task/task.c:769]: same comment here about if this is intended to be uncommented

File: iree/task/queue.h
Comment (ScottTodd) [inline iree/task/queue.h:105]: typo
```suggestion
// Future improvement idea: have the owner of the queue maintain a theft point
```

**Review by ScottTodd — CHANGES_REQUESTED** [review]

File: iree/base/math.h
Comment (benvanik) [reply iree/base/math.h:432]: Everything in here is thread-compatible - same with any RNG available in C or C++ (rand, std::mt19937, etc) - not sure it's worth calling out as it'd be weirder if any of this had anything to do with threading :)

File: iree/base/debugging.h
Comment (benvanik) [reply iree/base/debugging.h]: yeah copied too much from synchronization.h

File: iree/task/executor_test.cc
Comment (benvanik) [reply iree/task/executor_test.cc:106]: (will clean up this file a bit async of the rest of the review)

File: iree/base/math.h
Comment (benvanik) [reply iree/base/math.h:432]: Everything in here is thread-compatible - same with any RNG available in C or C++ (rand, std::mt19937, etc) - not sure it's worth calling out as it'd be weirder if any of this had anything to do with threading :)

**Review by ScottTodd — APPROVED** [review]
LGTM! If you're not in a hurry to unwind your stack of pending changes, maybe wait for @asaadaldien to get back from vacation and review since he expressed interest?

**Comment by benvanik** [comment]
> LGTM! If you're not in a hurry to unwind your stack of pending changes, maybe wait for @asaadaldien to get back from vacation and review since he expressed interest?

Want to get the HAL inversion stuff landed, and he's gone for a month, so he can post-review :)

**Pull request merged**
